core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java

```diff
@@ -96,27 +112,85 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeBoolean(this.isPrimary());
         out.writeOptionalString(this.getAssignedNodeId());
         out.writeOptionalWriteable(this.getUnassignedInfo());
+        out.writeVLong(remainingDelayNanos);
+        out.writeVInt(activeAllocationIds.size());
+        for (String id : activeAllocationIds) {
+            out.writeString(id);
+        }
 
-        Map<DiscoveryNode, Decision> ntd = this.getNodeDecisions();
-        out.writeVInt(ntd.size());
-        for (Map.Entry<DiscoveryNode, Decision> entry : ntd.entrySet()) {
-            entry.getKey().writeTo(out);
-            Decision.writeTo(entry.getValue(), out);
+        out.writeVInt(this.nodeExplanations.size());
+        for (NodeExplanation explanation : this.nodeExplanations.values()) {
+            explanation.writeTo(out);
         }
-        Map<DiscoveryNode, Float> ntw = this.getNodeWeights();
-        out.writeVInt(ntw.size());
-        for (Map.Entry<DiscoveryNode, Float> entry : ntw.entrySet()) {
-            entry.getKey().writeTo(out);
-            out.writeFloat(entry.getValue());
+    }
+
+    private NodeExplanation calculateNodeExplanation(DiscoveryNode node,
+                                                     Decision nodeDecision,
+                                                     Float nodeWeight,
+                                                     IndicesShardStoresResponse.StoreStatus storeStatus,
+                                                     String assignedNodeId,
+                                                     Set<String> activeAllocationIds) {
+        FinalDecision finalDecision;
+        StoreCopy storeCopy;
+        String finalExplanation;
+
+        if (node.getId().equals(assignedNodeId)) {
+            finalDecision = FinalDecision.ALREADY_ASSIGNED;
+            finalExplanation = "the shard is already assigned to this node";
```

Comment:
The example you gave in your last comment shows only single allocation decider same_shard as "NO" but "final_decision" is "NO" instead of "ALREADY_ASSIGNED". Is there an error in the example?